<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <title>NTM-DB: Browse by species</title>
    {% load static %}
    <link rel="icon" href="{%static '/image/logo/NTM_logo_light.svg'%}" />
    <link rel="stylesheet" type="text/css" href="{% static '/module/semantic-ui/semantic.min.css'%}">
    <link rel="stylesheet" type="text/css" href="{% static '/module/bootstrap/bootstrap.min.css'%}" />
    <link rel="stylesheet" type="text/css" href="{% static '/module/ngdc/page.min.css'%}" />
    <link rel="stylesheet" type="text/css" href="{% static '/css/supplementary.css'%}" />
    <link rel="stylesheet" type="text/css" href="/ntmdb/static/css/doc.css" />
    <style>
        .du-1 {
            margin: 0 0 3px 12px !important;
            font-size: medium !important;
        }

        .du-2 {
            margin: 0 0 3px 23px !important;
            font-size: medium !important;
        }

        .du-3 {
            margin: 0 0 3px 35px !important;
            font-size: medium !important;
        }

        .item-style {
            margin-bottom: -16px !important;
        }
    </style>
</head>


<body>
    {% include 'component/ngdc_header.html' %}
    {% include 'component/nav_bar2.html' %}
    <div id="main">
        <!-- breadcrumb -->
        <div class="ui basic segment">
            <div class="ui big breadcrumb" style="margin:8px 0 35px 0 !important;">
                <a href="\ntmdb" class="section"><i class="icon home"></i> Home</a>
                <i class="right angle icon divider"></i>
                <div class="active section">Documentation</div>
            </div>

            <div class="ui stackable grid row">
                <!-- semantic ui sidebar -->
                <div class="three wide column" style="margin:0 15px 0 25px !important; min-width: 247px;">
                    <div class="ui sticky" style="position: sticky !important; top: 50px !important;">
                        <div class="ui fluid secondary pointing vertical menu teal">
                            <div class="item item-style">
                                <a class="header bg-color sidebar-subtitle" href="#overview">Overview</a>
                            </div>
                            <div class="item item-style">
                                <a class="header bg-color sidebar-subtitle" href="#data-source">Data Source</a>
                            </div>
                            <div class="item item-style">
                                <a class="header bg-color sidebar-subtitle" href="#database_usage">Database Usage</a>
                                <div class="menu">
                                    <a class="item du-1" href="#browse">1. Browse</a>
                                    <a class="item du-2" href="#species">1.1 Species</a>
                                    <a class="item du-3" href="#species_page">1.1.1 Species Page</a>
                                    <a class="item du-2" href="#strains">1.2 Strains</a>
                                    <a class="item du-3" href="#strains_page">1.2.1 Strains Page</a>
                                    <a class="item du-3" href="#dst">1.2.2 DST</a>
                                    <a class="item du-2" href="#genes">1.3 Genes</a>
                                    <a class="item du-3" href="#genes_page">1.3.1 Genes Page</a>
                                    <a class="item du-1" href="#mlst">2. MLST</a>
                                    <a class="item du-1" href="#search">3. Search</a>
                                    <a class="item du-1" href="#analysis">4. Analysis</a>
                                    <a class="item du-1" href="#download">5. Download</a>
                                </div>
                            </div>
                            <div class="item item-style">
                                <a class="header bg-color sidebar-subtitle" href="#contact">Contact Us</a>
                            </div>
                        </div>
                    </div>
                </div>


                <div class="eleven wide column" style="margin-left:30px !important;">
                    <section id="overview" class="docTitle">
                        <!-- <img class="docuMark" src="/static/image/doc/doc_tmp_snow.png"> -->
                        <h2><i class="hand point right icon"></i> Overview</h2>
                        <p>The Non-Tuberculous Mycobacteria Database (NTM-DB) is a user-friendly genomic
                            database that
                            consolidates extensive genomic data, MLST genotyping, clinical phenotype
                            information, as well as
                            drug resistance and pathogenicity genotype information. This database offers
                            browsing,
                            searching, analysis, and download functionalities, aiming to provide clinicians,
                            researchers,
                            and drug developers with an opportunity for comprehensive comprehension of
                            non-tuberculous
                            mycobacterial infections, thereby fostering related research and enabling more
                            precise
                            therapies.</p>
                    </section>
                    <br>
                    <section id="data-source" class="docTitle">
                        <!-- <img class="docuMark" src="/static/image/doc/doc_tmp_snow.png"> -->
                        <h2><i class="hand point right icon"></i> Data Source</h2>
                        <p>NTM-DB collected all available sequenced NTM genomes
                            (totaling 16,083) and associate them with relevant metadata. Data sources include published
                            literature, publicly available sequenced strains in NCBI's advance release samples, as well
                            as strain assembly data collected from GenBank or RefSeq. The collected data encompass 178
                            species and 873 MLST types. </p>
                    </section>
                    <br>
                    <section id="database_usage" class="docTitle">
                        <!-- <img class="docuMark" src="/static/image/doc/doc_tmp_snow.png"> -->
                        <h2><i class="hand point right icon"></i> Database Usage</h2>
                        <br>

                        <!-- <h3 id="browse"> <i class="icon eye"></i> 1. Browse</h3> -->
                        <h3 id="browse">1. Browse</h3>
                        <p>The navigation bar features a "Browse" drop-down menu with three options:
                            "Species",
                            "Strains",
                            and "Genes". </p>

                        <h4 id="species"></i>1.1 Species</h4>
                        <p class="h4-p">By clicking on "Species", users can access information on the 190
                            non-tuberculous
                            mycobacteria
                            (NTM) species (including subspecies) collected in this database, including the
                            species name,
                            taxonomy ID, type strain genome <i
                                class="icon question circle outline qmark q-typeStrain"></i>,
                            representative genome <i class="icon question circle outline qmark q-represent"></i>, risk
                            group
                            <i class="icon question circle outline qmark q-riskGroup"></i>,
                            growth type <i class="icon question circle outline qmark q-growthType"></i>, as well as the
                            number of strains collected in NTM-DB that belong to each species.
                        </p>

                        <h5 id="species_page"></i>1.1.1 Species Page</h5>
                        <p class="h5-p">By clicking on the species name (<i>e.g.</i> <i>Mycobacterium
                                abscessus</i>), users
                            can access
                            the corresponding species detail
                            page to view
                            additional information, including phylogeny, multilocus sequence typing (MLST),
                            pan-genome,
                            pan-resistom, and pan-virolom.</p>

                        <h4 id="strains"></i>1.2 Strains</h4>
                        <p class="h4-p">In the "Strains" table, users can view basic information on the 16,083 strains
                            collected in this
                            database, including accession, taxonomy, representative strain, sequence typing,
                            isolation
                            source, identified drug resistance and virulence genes, along with their
                            corresponding
                            mechanisms. Additionally, whether there are any drug susceptibility testing
                            (DST)
                            results
                            available is also indicated. Furthermore, users can select details such as
                            collection time,
                            country, sequencing quality, and other information by checking the corresponding
                            boxes located
                            at the top of the page.</p>

                        <h5 id="strains_page"></i>1.2.1 Strains Page</h5>
                        <p class="h5-p">By clicking on the accession number (<i>e.g.</i> GCF_900454025.1),
                            users
                            can access
                            the corresponding species detail
                            page to view
                            additional information, including resistom and virolom.</p>

                        <h5 id="dst"></i>1.2.2 Drug Susceptibility Testing
                            (DST)</h5>
                        <p class="h5-p">To view all DST results stored in this database, users can click the
                            "View drug
                            susceptibility
                            testing (DST) results" button located above the table. The DST page provides
                            information on drug
                            resistance phenotypes for the 598 NTM strains included in our database. Users
                            can
                            select which
                            drug classes to display. In each cell, <span style="color: #DC143C; font-weight: bold;">[R]
                                represents "Resistant"</span>, <span style="color: #38B000; font-weight: bold;">[S]
                                represents "Sensitive"</span>, <span style="color: #FF8C00; font-weight: bold;">[I]
                                represents "Intermediate"</span>, numerical
                            values represent minimum inhibitory concentration (MIC), and "-" indicates that
                            corresponding
                            information was not provided in the original literature.
                        </p>

                        <h4 id="genes"></i>1.3 Genes</h4>
                        <p class="h4-p">The genes module includes two parts: drug resistance genes and virulence genes.
                            This
                            database
                            contains 154 drug resistance genes and 462 virulence genes, providing
                            information
                            such as gene
                            name, accession, related mechanisms and classifications for drug resistance and
                            virulence, as
                            well as the number of NTM species or strains in which specific genes exist.</p>

                        <h5 id="genes_page"></i>1.3.1 Genes Page</h5>
                        <p class="h5-p">Clicking on a gene name provides access to detailed information
                            about
                            the gene, as
                            well as its
                            prevalence in sequenced NTM genomes. Furthermore, we extract the region where
                            the
                            reference
                            sequence of the gene hits from one strain of each species.</p>


                        <!-- <h3 id="mlst"><i class="icon sun outline"></i> 2. Multilocus Sequence Typing (MLST)</h3> -->
                        <h3 id="mlst">2. Multilocus Sequence Typing (MLST)</h3>
                        <p>As sequence typing is of significant importance for non-tuberculous mycobacteria,
                            the
                            NTM-DB
                            provides users with a module specifically designed for viewing the sequence
                            types
                            (STs) of all
                            NTM species included in the database. Users can choose between "Colored by ST"
                            or
                            "Colored by
                            Species" options. When hovering over a circle representing a particular ST (e.g.
                            ST8), detailed
                            information about that ST will be displayed. Clicking on a particular ST will
                            take
                            the user to
                            the [Species Page] and highlight that ST on the MLST diagram.</p>


                        <!-- <h3 id="search"><i class="icon search"></i>3. Search</h3> -->
                        <h3 id="search">3. Search</h3>
                        <p>NTM-DB offers three advanced search modules on the search page, which are as follows:
                        </p>
                        <p>(1) The "Species" module enables users to input a species name (<i>e.g.</i>
                            <i>M.abscessus</i>)
                            and match NTM species to obtain detailed information on the [Species Page].
                        </p>
                        <p>(2) Similarly, under the "Strains" module, users can enter the accession number of
                            a
                            strain
                            (<i>e.g.</i> GCF_001077755.1) to retrieve its corresponding page [Strains Page]
                            and
                            view
                            detailed information about the strain.</p>
                        <p>(3) The "Genes" module is divided into drug resistance genes and virulence genes.
                            Users can also
                            enter a gene name (<i>e.g.</i> rpoB) to access its corresponding page [Genes
                            Page]
                            and view
                            information about the gene. Furthermore, users can filter genes based on criteria
                            such
                            as
                            resistance to specific drug class or virulence factor category (VFC).</p>
                        <p>* It is important to note that all search boxes are mutually exclusive.</p>



                        <!-- <h3 id="analysis"><i class="icon calendar check outline"></i>4. Analysis</h3> -->
                        <h3 id="analysis">4. Analysis</h3>
                        <p>The Analysis page provides users with personalized analysis, including building
                            evolutionary
                            trees, MLST identification, and annotation of drug resistance genes and
                            virulence
                            genes. First,
                            the user needs to submit a strain sequence file that meets the required format
                            in
                            the form on
                            the page, and then select the desired analysis items. It should be noted that we
                            provide users
                            with two ways to build evolutionary trees:</p>
                        <p>(1) Selecting the number of closest nodes (representative strain of each closest
                            species).</p>
                        <p>(2) Manually selecting the names of the desired species (multiple choices).</p>
                        <p>After making the selection and clicking submit button, the system will begin the
                            analysis and
                            display the evolutionary tree on the analysis results page, along with download
                            links for drug
                            resistance and virulence annotations. Users can download these files as needed.
                        </p>

                        <!-- <h3 id="download"> <i class="icon download"></i> 5. Download</h3> -->
                        <h3 id="download">5. Download</h3>
                        <p>The original data, relevant information regarding virulence and drug resistance
                            genes, MLST of
                            each species, and personalized analysis results of NTM strains in NTM-DB are all
                            available for
                            download.</p>
                    </section>
                    <br>
                    <section id="contact" class="docTitle">
                        <!-- <img class="docuMark" src="/static/image/doc/doc_tmp_snow.png"> -->
                        <h2><i class="hand point right icon"></i> Contact Us</h2>
                        <p>If you have any questions, suggestions or feedback about NTM-DB, please do not
                            hesitate to
                            contact us. Our team is dedicated to providing excellent service and support,
                            and we
                            will get
                            back to you as soon as possible. You can reach us by emailing us directly at [
                            <span style="color: #3391b9;"> ntmdb@big.ac.cn </span> ].
                            Thank you for your interest in NTM-DB, and we look forward to hearing from you.
                        </p>
                    </section>
                    <br>
                </div>
            </div>
        </div>
    </div>


    {% include 'component/ngdc_footer.html' %}
    <script src="{% static '/module/jquery/jquery-3.6.4.js'%}"></script>
    <script src="{% static '/module/bootstrap/bootstrap.bundle.min.js'%}"></script>
    <script src="{% static '/module/semantic-ui/semantic.min.js'%}"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/metisMenu/3.0.7/metisMenu.min.js"></script>
    <script src="{% static 'js/doc/doc.js' %}" type="text/javascript"></script>
    <script>
        $('.question.circle.outline.q-typeStrain').popup({
            content: '[Definition from LPSN] Type strain serves as the reference for defining and describing that particular species. It is typically the first strain of the species to be discovered, well-characterized, and deposited in a recognized culture collection. The type strain serves as a reference point for future studies and comparisons within that species.'
        });
        $('.question.circle.outline.q-represent').popup({
            content: '[Definition from NCBI] A complete genome sequence selected as a representative for a defined species, which lacks a reference genome. The selection process takes into consideration several criteria such as assembly quality, completeness, and gene count.'
        });
        $('.question.circle.outline.q-riskGroup').popup({
            content: '[Risk Group 0]: Taxonomic units mentioned in TRBA 466 but not categorized into specific risk groups.  [Risk Group 1]: Unlikely to cause disease in humans or animals and pose a low risk of infection.  [Risk Group 2]: Can cause disease in humans or animals but can be safely handled with appropriate precautions.  [Risk Group 3]: Associated with serious or lethal disease in humans or animals, or with ecological harm, and require highly skilled expertise, advanced laboratory equipment, and safe working environments for handling.'
        });
        $('.question.circle.outline.q-growthType').popup({
            content: '[Rapidly growing]: form visible colonies on agar medium in less than 7 days. [Slowly growing]: form visible colonies on agar medium only after 7 days.'
        });



    </script>
</body>

</html>